### PR TITLE
fix error when first argument of replace function is const

### LINKED ISSUE #9522: Run sql failed when using replace function
## Bug Report

When using the `replace` function in complex SQL, it may be affected by the `where` condition, causing the `replace` function to fail when pushed down to Tiflash for execution.
For example, 
```SQL
SELECT /*+ read_from_storage(tiflash[my_table]) */ REPLACE(my_table.col14, my_table.col_11, my_table.col_14) FROM my_table where my_table.col14 in 'TEST SQL';
```
This statement will replace the `replace` function's parameter with the filter parameter when it is pushed down to improve SQL execution efficiency. Unfortunately, Tiflash only supports the `replace` function first parameter to be of type `ColumnString`.

### 1. Minimal reproduce step (Required)
```SQL
mysql> CREATE TABLE my_table (
         `col_11` VARCHAR(255) NOT NULL,
         `col_12` int(10) unsigned NOT NULL,
         `col_13` VARCHAR(255) NOT NULL,
         `col_14` VARCHAR(255) NOT NULL
     );
Query OK, 0 rows affected (0.02 sec)

mysql> insert into my_table values ('World',1203945,'Hello, World!','FOR TEST'), ('ttttt',1122,'bbbbb','cccc');
Query OK, 1 row affected (0.00 sec)

mysql> SELECT      /*+ read_from_storage(tiflash[my_table]) */      REPLACE(my_table.col_13, my_table.col_11, my_table.col_13)  FROM      my_table;
+------------------------------------------------------------+
| REPLACE(my_table.col_13, my_table.col_11, my_table.col_13) |
+------------------------------------------------------------+
| Hello, Hello, World!!                                      |
+------------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select * from my_table;
+--------+---------+---------------+----------+
| col_11 | col_12  | col_13        | col_14   |
+--------+---------+---------------+----------+
| World  | 1203945 | Hello, World! | FOR TEST |
| ttttt  |    1122 | bbbbb         | cccc     |
+--------+---------+---------------+----------+
2 rows in set (0.00 sec)

mysql> select replace('Hello World',my_table.col_11, my_table.col_14) from my_table;
+---------------------------------------------------------+
| replace('Hello World',my_table.col_11, my_table.col_14) |
+---------------------------------------------------------+
| Hello FOR TEST                                          |
| Hello World                                             |
+---------------------------------------------------------+
2 rows in set (0.00 sec)

mysql> explain select replace('Hello World',my_table.col_11, my_table.col_14) from my_table;
+-------------------------+---------+-----------+----------------+----------------------------------------------------------------------------+
| id                      | estRows | task      | access object  | operator info                                                              |
+-------------------------+---------+-----------+----------------+----------------------------------------------------------------------------+
| Projection_3            | 2.00    | root      |                | replace(Hello World, test.my_table.col_11, test.my_table.col_14)->Column#6 |
| └─TableReader_6         | 2.00    | root      |                | data:TableFullScan_5                                                       |
|   └─TableFullScan_5     | 2.00    | cop[tikv] | table:my_table | keep order:false, stats:pseudo                                             |
+-------------------------+---------+-----------+----------------+----------------------------------------------------------------------------+
3 rows in set (0.00 sec)

mysql> ALTER TABLE my_table SET TIFLASH REPLICA 1;
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT /*+ read_from_storage(tiflash[my_table]) */ REPLACE('Hello World', my_table.col_11, my_table.col_14)  FROM my_table;
ERROR 1105 (HY000): other error for mpp stream: Code: 44, e.displayText() = DB::Exception: Illegal column Const(String) of first argument of function replaceAll, e.what() = DB::Exception,



mysql> explain SELECT /*+ read_from_storage(tiflash[my_table]) */     REPLACE('Hello World', my_table.col_11, my_table.col_14)  FROM my_table;
+---------------------------+---------+--------------+----------------+----------------------------------------------------------------------------+
| id                        | estRows | task         | access object  | operator info                                                              |
+---------------------------+---------+--------------+----------------+----------------------------------------------------------------------------+
| TableReader_10            | 2.00    | root         |                | MppVersion: 2, data:ExchangeSender_9                                       |
| └─ExchangeSender_9        | 2.00    | mpp[tiflash] |                | ExchangeType: PassThrough                                                  |
|   └─Projection_4          | 2.00    | mpp[tiflash] |                | replace(Hello World, test.my_table.col_11, test.my_table.col_14)->Column#6 |
|     └─TableFullScan_8     | 2.00    | mpp[tiflash] | table:my_table | keep order:false, stats:pseudo                                             |
+---------------------------+---------+--------------+----------------+----------------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

### 2. What did you expect to see? (Required)
Query OK.
### 3. What did you see instead (Required)
ERROR 1105 (HY000): other error for mpp stream: Code: 44, e.displayText() = DB::Exception: Illegal column Const(String) of first argument of function replaceAll, e.what() = DB::Exception,
### 4. What is your TiFlash version? (Required)
master

### PR DESCRIPTION
**Pull request by guo-shaoge**
### What problem does this PR solve?

Issue Number: close #9522

Problem Summary:

### What is changed and how it works?

```commit-message
1. remove some useless arguments(pos, occ, match_type)
2. support first argument as ColumnConst
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
fix error when first argument of replace function is const
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsStringReplace.h [hunk 1]
```diff
@@ -217,13 +153,76 @@ class FunctionStringReplace : public IFunction
...
+        bool needle_const,
+        bool replacement_const,
+        ColumnWithTypeAndName & column_result) const
+    {
+        auto res_col = ColumnString::create();
+        res_col->reserve(column_src->size());
+
+        RUNTIME_CHECK_MSG(
+            !needle_const || !replacement_const,
+            "should got here when all argments of replace is constant");
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsStringReplace.h:169]: you mean "should not got here when all arguments of replace is constant"?

File: dbms/src/Functions/FunctionsStringSearch.cpp [hunk 2]
```diff
@@ -816,9 +816,6 @@ struct ReplaceStringImpl
         const ColumnString::Offsets & offsets,
         const std::string & needle,
         const std::string & replacement,
-        const Int64 & /* pos */,
-        const Int64 & /* occ */,
-        const std::string & /* match_type */,
         TiDB::TiDBCollatorPtr /* collator */,
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsStringSearch.cpp:819]: seems `collator` is also not used.

File: dbms/src/Functions/FunctionsStringReplace.h (hunk 1 above)
Comment (guo-shaoge) [reply dbms/src/Functions/FunctionsStringReplace.h:169]: yes typo

File: dbms/src/Functions/FunctionsStringSearch.cpp (hunk 2 above)
Comment (guo-shaoge) [reply dbms/src/Functions/FunctionsStringSearch.cpp:819]: done

File: tests/fullstack-test/expr/replace.test [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2024 PingCAP, Inc.
```
Comment (windtalker) [inline tests/fullstack-test/expr/replace.test:1]: add some ut for this?

**Review by yibin87 — APPROVED** [review]
Add uts, others LGTM

**Review by windtalker — APPROVED** [review]
lgtm

**Pull request merged**